README.md
# Stokvel Coin
Stokvel Coin is a crypto currency based on EOS blockchain, 

# What is a Stokvel

A Stokvel is an investment club. Usually you become a member by invitation-only, members of a Stokvel often belong to a close knit community. Stokvels are formally defined as a type of a investment club in which a group of people enter into an agreement to contribute a fixed amount of money to a common pool either weekly, fortnightly or monthly. A Stokvel serves as a rotating credit union or saving schemes where members contribute fixed sums of money to a central fund on a particular basis. Stokvels are not uniquely a South African phenomenon, you get them in various forms, such as Burial Societies, Investment Clubs, Umgalelo Clubs, Mavuso Stokvel and Youth Stokvels. Stokvels are not regulated by the South African financial regulators unlike Banks and Insurance companies, hence they are not thought of as part of the formal financial business sector. Stokvel are governed by agreed upon rules and regulations that are part of its constitution(founding document), there is no legal protection offered by financial regulators.

There are a number of pain points in the Stokvel Business Model.

* The periodic(weekly, fortnightly or monthly) contributions are compulsory and skipping contributions can tantamount to membership cancellation.
* The periodic contributions often have a hard cutoff date that no member is allowed to miss, missing a payment tantamount to disqualification. This is a huge problem that invariably lead to conflicts between members.
* Some Stokvels don't allow members to join after the commencement date, the commencement date is usually the first month of the year.
* Barriers to Digital transformation are prevalent in the informal economy.

This research project is geared to investigate how the Stokvel Business model can be implemented using Blockchains, Ricardian and Smart Contracts.



## Software Installation
### Below are instructions for downloading and installing the software needed.
  * sudo yum update
  * wget https://github.com/eosio/eos/releases/download/v2.2.0-rc1/eosio-2.2.0-rc1.el7.x86_64.rpm
  * sudo yum install ./eosio-2.2.0-rc1.el7.x86_64.rpm
  * wget https://github.com/EOSIO/eosio.cdt/releases/download/v1.8.1/eosio.cdt-1.8.1-1.amzn.x86_64.rpm
  * sudo yum install $eosio.cdt-1.8.1-1.amzn.x86_64.rpm

### Install the Stokfel Token and Contract.
  * sudo mkdir -p /opt/stokvel-coin/code
  * sudo chown ec2-user:ec2-user -R /opt/stokvel-coin/
  * sudo yum install -y git
  * cd /opt/stokvel-coin/code
  * git clone https://github.com/phumlani-mbabela/stokvel-coin.git
  * sudo mkdir -p /opt/stokvel-coin/code/stokvel-coin/secrets/
  * sudo chown ec2-user:ec2-user -R /opt/stokvel-coin/
  
### Create a service startup script[Optional].This file is available in the code repo.
  * sudo vi /etc/systemd/system/eosio.service
  * ...  ...
  * sudo systemctl daemon-reload
  * sudo systemctl enable eosio.service
  * sudo systemctl reboot

## Blockchain Testing
### Below are instructions for starting and testing the blockchain.
  * Start server - sudo nodeos -e -p eosio --p2p-peer-address=0.0.0.0:9876 --plugin eosio::net_api_plugin --plugin eosio::net_plugin --plugin eosio::producer_plugin --plugin eosio::producer_api_plugin --plugin eosio::chain_api_plugin --plugin eosio::http_plugin --plugin eosio::history_plugin --plugin eosio::history_api_plugin --filter-on="*" --access-control-allow-origin='*' --contracts-console --http-validate-host=false --verbose-http-errors --replay-blockchain  > /var/log/eosio-startup.out 2>&1 
  * View the blockchain process - sudo ps -eaf | grep nodeos
  * View logs - tail -f /var/log/eosio-startup.out
  *  Protocol Features - curl --request POST --url http://127.0.0.1:8888/v1/producer/schedule_protocol_feature_activations -d '{"protocol_features_to_activate": ["0ec7e080177b2c02b278d5088611686b49d739925a92d9bfcacd7fc6b74053bd"]}'
  * Get system information - cleos --url http://13.245.150.240:8888  get info


## Accounts Configuration
### Setup Wallets
* cleos wallet create -n w_admin -f /opt/stokvel-coin/code/stokvel-coin/secrets/w_admin.pwd   
* cleos wallet create -n w_anele -f /opt/stokvel-coin/code/stokvel-coin/secrets/w_anele.pwd   
* cleos wallet create -n w_alice -f /opt/stokvel-coin/code/stokvel-coin/secrets/w_alice.pwd   
* cleos wallet create -n w_thabo -f /opt/stokvel-coin/code/stokvel-coin/secrets/w_thabo.pwd   
* cleos wallet create -n w_james -f /opt/stokvel-coin/code/stokvel-coin/secrets/w_james.pwd   
* cleos wallet create            -f /opt/stokvel-coin/code/stokvel-coin/secrets/w_default.pwd 

### Create keys
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_admin.txt 
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_admin1.txt 
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_anele.txt 
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_alice.txt 
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_thabo.txt 
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_james.txt 
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_default.txt 
* cleos create key --file /opt/stokvel-coin/code/stokvel-coin/secrets/keys_default1.txt 

### Import keys to the corresponding wallets.
* cleos wallet open 
* cleos wallet unlock --password PW5JzxxNgSGEubmxBT6hiWZbtURpL6kkfyfgkjY7n36XAYRdqjc8Z 
* cleos wallet import --private-key 5J6jJNEi7x8uH77kUBuZmcVmSZ9vDustnCigDcAv9E7ewzmHvf1 
* cleos wallet import --private-key 5J4ESvWknUcKsdFQfBDcY7j5hD5FED7r26zFod8fGBPrwT3Cqwh 

* cleos wallet open -n w_admin 
* cleos wallet unlock -n w_admin --password PW5Jf5J9drtA39FMgJ2FqrKikudxY6XVFD9YebyFGhhMCb2Kqbyus 
* cleos wallet import --name w_admin --private-key 5JSxZAGFnyT9rfUbnZBRLWrhYSUssjyZLRHJnnCoqSogPuq89C9 
* cleos wallet import --name w_admin --private-key 5JSxZAGFnyT9rfUbnZBRLWrhYSUssjyZLRHJnnCoqSogPuq89C9 

* cleos wallet open -n w_anele 
* cleos wallet unlock -n w_anele --password PW5K1i3ka9ctFpX8n1ADHojtA7CZtd1pYpJPry6UNQ7udKLxNooEM 
* cleos wallet import --name w_anele --private-key 5JTmMUrfc9iujY7mmh3LBjZCwB88qxoiM5qiHnipgG6DPww5G8Q 

* cleos wallet open -n w_alice 
* cleos wallet unlock -n w_alice --password PW5KHdAMqNREGnrZPnggi8FFADgFaR4NTbf8UBjxdWxsk3CsEzQM5 
* cleos wallet import --name w_alice --private-key 5J2g8LarmhHd2PjyWzTEVaZiXDPqwBS3Dg3VJmWQyTnJ5vhv8CB 

* cleos wallet open -n w_thabo 
* cleos wallet unlock -n w_thabo --password PW5KG3EvAvosExvADsYyRpHghJTEKNbE7dgQVPY8RjLEGDD54Zjbu 
* cleos wallet import --name w_thabo --private-key 5HwXyKDF7PMF9VSj7JJJ74JRAzR5NaUmfxJScBXm5LxU4sez7Ck 

* cleos wallet open -n w_james
* cleos wallet unlock -n w_james --password PW5JTVwsPJCiLp5fY9HGM5HEPcmw6J6t9PU7J123XgKb1cPgvtKHp
* cleos wallet import --name w_james --private-key 5KC5Mu4pK1ewDPHALdG99n8d8DWufMQnaaabws6hSmYroLopKmX

### Import the eosio default private key.
* cleos wallet import                --private-key 5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD  
* cleos wallet import --name w_admin --private-key 5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3 
* cleos wallet import --name w_anele --private-key 5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3 
* cleos wallet import --name w_alice --private-key 5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3 
* cleos wallet import --name w_thabo --private-key 5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3 
* cleos wallet import --name w_james --private-key 5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3 


### Create accounts
* cleos create account eosio accadmin EOS78EE82iFKgZtg6PdY9wv8R5Jp19x1oEf931NR2A2B75yJYJjnm EOS62PqFepS9PwH41wwYLCK9pxTj6vbRbUW6EY8EA56abudxZdayB -p eosio@active
* cleos create account eosio accanele EOS8QR2MvXR6vRgeACNX2KRc5iToRdm7ZAjXFvgAPxZ1z5FSNrepa EOS8QR2MvXR6vRgeACNX2KRc5iToRdm7ZAjXFvgAPxZ1z5FSNrepa -p eosio@active 
* cleos create account eosio accalice EOS8WL6ipyQWGFo3vAHE43SoHArj9mwJ2j3XRVRUrhcx3U8eFU9n4 EOS8WL6ipyQWGFo3vAHE43SoHArj9mwJ2j3XRVRUrhcx3U8eFU9n4 -p eosio@active 
* cleos create account eosio accthabo EOS8MV5EqvYJXNBKRkEE5Yp3KLQdQwnFA63ZZLADJDHnz7D4GkaTj EOS8MV5EqvYJXNBKRkEE5Yp3KLQdQwnFA63ZZLADJDHnz7D4GkaTj -p eosio@active  
* cleos create account eosio accjames EOS5mcL66PowDxrv9TypzXkQsnoonJEnEdFJUK1Mg94WK9k5xExkj EOS5mcL66PowDxrv9TypzXkQsnoonJEnEdFJUK1Mg94WK9k5xExkj -p eosio@active 

### Test Accounts
* cleos get account accadmin --json
* cleos get account accanele --json
* cleos get account accalice --json
* cleos get account accthabo --json
* cleos get account accjames --json


## Compile and Deploy the code

### Compile the token contract
* cd /opt/stokvel-coin/code/stokvel-coin/eosiostokveltoken/build
* eosio-cpp -abigen ../src/eosiostokveltoken.cpp  -o eosiostokveltoken.wasm -I ../include/
* cd /opt/stokvel-coin/code/stokvel-coin/bank/build
* eosio-cpp -abigen ../src/bank.cpp  -o bank.wasm -I ../include/
  
### Deploy the token contract
* cd /opt/stokvel-coin/code/stokvel-coin/bank/build
* cleos set contract accadmin . --abi eosiostokveltoken.abi eosiostokveltoken.wasm -p accadmin@active

## Create a new token
To create the STVL token use the following command:
The create action mints 21 million tokens.
    * cleos push action accadmin createe '["accadmin","210000000000.0000 STVL", "created 210000000000.0000 STVL"]' -p accadmin@active
### Alternative invocation
    * cleos push action accadmin create '[]' -p accadmin@active
    
### View the newly created token.
    * cleos get table accadmin STVL stat        
    * cleos get currency stats accadmin STVL

## Issue New Tokens
### Issue 100 STVL tokens. To execute the action, use the following command:
    * cleos push action accadmin issue '["210000000000.0000 STVL", "issue 210000000000 STVL"]' -p accadmin@active
    * cleos push action accadmin issuee '["accadmin", "210000000000.0000 STVL", "issued 210000000000 STVL"]' -p accadmin@active
    * cleos get table accadmin STVL stat
          
## Transfer tokens from the suspense account to a different account.
    * cleos push action accadmin transfer '[ "accadmin", "accanele", "101.0101 STVL", "Transfer 101 STVL tokens from suspense account." ]' -p accadmin@active
    * cleos push action accadmin transfer '[ "accadmin", "accalice", "102.0202 STVL", "Transfer 102 STVL tokens from suspense account." ]' -p accadmin@active
    * cleos push action accadmin transfer '[ "accadmin", "accthabo", "103.0303 STVL", "Transfer 103 STVL tokens from suspense account." ]' -p accadmin@active
    * cleos push action accadmin transfer '[ "accadmin", "accjames", "104.0404 STVL", "Transfer 104 STVL tokens from suspense account." ]' -p accadmin@active
### Alternative invocation.
    * cleos transfer accadmin accjames "202.0202 STVL" "Transfer 202 STVL tokens from suspense account." -c accadmin -p accadmin

## Inspect balances.
    * cleos get currency balance accadmin accadmin STVL
    * cleos get currency balance accadmin accanele STVL
    * cleos get currency balance accadmin accalice STVL
    * cleos get currency balance accadmin accthabo STVL
    * cleos get currency balance accadmin accjames STVL
### Alternative invocation.
    * cleos get account accadmin --json
    * cleos get account accanele --json
    * cleos get account accalice --json
    * cleos get account accthabo --json
    * cleos get account accjames --json

## Airgrab tokens
    * cleos push action accadmin airgrab '[ "accanele"]' -p accanele@active
    * cleos push action accadmin airgrab '[ "accanele", "100.0000 STVL"]' -p accanele@active
    * cleos get currency balance accadmin accanele STVL

##### Project Founder: Phumlani Kaida Mbabela

history.txt
    1  wget https://github.com/eosio/eos/releases/download/v2.2.0-rc1/eosio-2.2.0-rc1.el7.x86_64.rpm
    2  sudo yum install ./eosio-2.2.0-rc1.el7.x86_64.rpm
    3  sudo yum update
    4  nodeos -e -p eosio --plugin eosio::producer_plugin --plugin eosio::producer_api_plugin --plugin eosio::chain_api_plugin --plugin eosio::http_plugin --plugin eosio::history_plugin --plugin eosio::history_api_plugin --filter-on="*" --access-control-allow-origin='*' --contracts-console --http-validate-host=false --verbose-http-errors >> nodeos.log 2>&1 
    5  sudo ps -eaf | grep eos
    6  sudo ps -eaf | grep eosio
    7  sudo kill -9 5893 5875  5874  5873  5889 5875 5874 5873 
    8  sudo ps -eaf | grep eosio
    9  nodeos -e -p eosio --plugin eosio::producer_plugin --plugin eosio::producer_api_plugin --plugin eosio::chain_api_plugin --plugin eosio::http_plugin --plugin eosio::history_plugin --plugin eosio::history_api_plugin --filter-on="*" --access-control-allow-origin='*' --contracts-console --http-validate-host=false --verbose-http-errors >> nodeos.log 2>&1 
   10  sudo vi /var/tmp/eosio-startup.sh
   11  sudo chmod +x /var/tmp/eosio-startup.sh
   12  vi /etc/systemd/system/eosio.service
   13  sudo vi /etc/systemd/system/eosio.service
   14  sudo systemctl daemon-reload
   15  sudo systemctl enable eosio.service
   16  systemctl reboot
   17  sudo systemctl reboot
   18  vi /var/tmp/eosio-startup.sh
   19  clear
   20  tail -f /var/log/eosio-startup.out
   21  pwd
   22  ls -lst
   23  pwd
   24  ls -lst /
   25  cd /opt/
   26  ls
   27  mkdir stokvel-coin
   28  sudo mkdir stokvel-coin
   29  ls -lst
   30  cd stokvel-coin/
   31  ls -lst
   32  sudo mkdir code
   33  cd code/
   34  ls
   35  pwd
   36  git clone https://github.com/phumlani-mbabela/stokvel-coin.git
   37  sudo yum install git
   38  sudo yum install -y git
   39  git clone https://github.com/phumlani-mbabela/stokvel-coin.git
   40  sudo git clone https://github.com/phumlani-mbabela/stokvel-coin.git
   41  ls
   42  cd stokvel-coin/
   43  ls
   44  ls -lst
   45  mkdir build
   46  sudo mkdir build
   47  pwd
   48  sudo chown ec2-user:ec2-user -R /opt/stokvel-coin/
   49  pwd
   50  mkdir build
   51  cd build/
   52  ls
   53  pwd
   54  ls ../src/
   55  eosio-cpp -abigen ../src/bank.cpp  -o bank.wasm -I ../include/
   56  wget https://github.com/eosio/eosio.cdt/releases/download/v1.8.0/eosio.cdt-1.8.0-1.el7.x86_64.rpm
   57  wget https://github.com/EOSIO/eosio.cdt/releases/download/v1.8.1/eosio.cdt-1.8.1-1.amzn.x86_64.rpm
   58  sudo yum install ./eosio.cdt-1.8.1-1.amzn.x86_64.rpm 
   59  sudo yum install ./eosio.cdt-1.8.1-1.amzn.x86_64.rpm -y
   60  eosio-cpp -abigen ../src/bank.cpp  -o bank.wasm -I ../include/
   61  pwd
   62  cd ..
   63  ls -lst
   64  eosio-init --project=helloworld
   65  ls -0lst
   66  ls -lst
   67  cd helloworld/
   68  ls
   69  pwd
   70  cd ..
   71  ls -lst
   72  pwd
   73  ls -lst
   74  pwd
   75  eosio-init --project=bank
   76  eosio-init --project=eosio.stokvel.token
   77  eosio-init --project=eosio-stokvel-token
   78  eosio-init --project=eosiostokveltoken
   79  ls -lst
   80  rm helloworld/
   81  rm -r  helloworld/
   82  ls -lst
   83  git add .
   84  git  commit -m "Added contracts via eosio-init --project="
   85  git push
   86  pwd
   87  ls
   88  mkdir certs
   89  cd certs
   90  ssh-keygen -t ed25519 -C "pmbabela@gmail.com"
   91  pwd
   92  ls
   93  pwd
   94  sudo cp /home/ec2-user/.ssh/id_ed25519-pmbabela@gmail.com  /opt/stokvel-coin/code/stokvel-coin/certs/id_ed25519-pmbabela@gmail.com
   95  sudo cp /home/ec2-user/.ssh/id_ed25519-pmbabela@gmail.com.pub /opt/stokvel-coin/code/stokvel-coin/certs/id_ed25519-pmbabela@gmail.com.pub
   96  pwd
   97  cd ..
   98  git add .
   99  pwd
  100  sudo chown ec2-user:ec2-user -R /opt/stokvel-coin/
  101  git add .
  102  pwd/opt/stokvel-coin/code/stokvel-coin/certs/id_ed25519-pmbabela@gmail.com.pub
  103  vi /opt/stokvel-coin/code/stokvel-coin/certs/id_ed25519-pmbabela@gmail.com.pub
  104  https://www.gnupg.org/ftp/gcrypt/gnupg/gnupg-2.3.4.tar.bz2
  105  wget https://www.gnupg.org/ftp/gcrypt/gnupg/gnupg-2.3.4.tar.bz2
  106  tar -xvjf gnupg-2.3.4.tar.bz2 
  107  ls
  108  gpg 
  109  cd certs/
  110  ls 
  111  gpg --full-generate-key
  112  gpg --default-new-key-algo rsa4096 --gen-key
  113  history
  114  gpg --default-new-key-algo rsa4096 --gen-key
  115  pwd
  116  mkdir /opt/stokvel-coin/code/stokvel-coin/certs/pgp
  117  pwd
  118  ls -lst
  119  cd ..
  120  ls -lst
  121  pwd
  122  cd ..
  123  ls
  124  mv -r stokvel-coin/ stokvel-coin-/
  125  mv  stokvel-coin/ stokvel-coin-/
  126  ls -lst
  127  git clone phumlani-mbabela@github.com:phumlani-mbabela/stokvel-coin.git
  128  git clone git@github.com:phumlani-mbabela/stokvel-coin.git
  129  ssh-agent bash -c 'ssh-add /opt/stokvel-coin/code/stokvel-coin-/certs/id_ed25519-pmbabela@gmail.com; git clone git@github.com:phumlani-mbabela/stokvel-coin.git'
  130  ls -lst
  131  cd stokvel-coin
  132  ls 
  133  eosio-init --project=bank
  134  eosio-init --project=eosiostokveltoken
  135  ls 
  136  ls -lst
  137  cd bank/
  138  ls
  139  cd include/
  140  ls
  141  cd ..
  142  ls -lst
  143  git add .
  144  git commit -m "Added 2 contracts"
  145  git push
  146  eval "$(ssh-agent -s)"
  147  vi ~/.ssh/config
  148  ls ~/.ssh/
  149  ls -ls ~/.ssh/
  150  vi ~/.ssh/config
  151  ssh-add -K ~/.ssh/id_ed25519-pmbabela@gmail.com
  152  ssh-add -k ~/.ssh/id_ed25519-pmbabela@gmail.com
  153  eval "$(ssh-agent -s)"
  154  ssh -T git@github.com
  155  ssh -T ec2-user@github.com
  156  ls -lst /home/ec2-user/.ssh/config
  157  ls -lst /home/ec2-user/.ssh/config/
  158  ssh -T git@github.com
  159  chmod 600 ~/.ssh/config
  160  ssh -T git@github.com
  161  vi ~/.ssh/config
  162  ssh -T git@github.com
  163  pwd
  164  ls -lst
  165  git push
  166  pwd
  167  ls -lst
  168  cd bank/
  169  ls -lst
  170  ls src/
  171  cd build/
  172  ls

eosiostokveltoken/include/eosiostokveltoken.hpp
#pragma once

#include <eosio/asset.hpp>
#include <eosio/eosio.hpp>

#include <string>

namespace eosiosystem {
	class system_contract;
}

namespace eosio {

	using std::string;

	/**
	 * The `eosiostokveltoken` sample system contract defines the structures and actions that allow users to create, issue, and manage tokens for EOSIO based blockchains. It demonstrates one way to implement a smart contract which allows for creation and management of tokens. It is possible for one to create a similar contract which suits different needs. However, it is recommended that if one only needs a token with the below listed actions, that one uses the `eosio.token` contract instead of developing their own.
	 *
	 * The `eosiostokveltoken` contract class also implements two useful public static methods: `get_supply` and `get_balance`. The first allows one to check the total supply of a specified token, created by an account and the second allows one to check the balance of a token for a specified account (the token creator account has to be specified as well).
	 *
	 * The `eosiostokveltoken` contract manages the set of tokens, accounts and their corresponding balances, by using two internal multi-index structures: the `accounts` and `stats`. The `accounts` multi-index table holds, for each row, instances of `account` object and the `account` object holds information about the balance of one token. The `accounts` table is scoped to an EOSIO account, and it keeps the rows indexed based on the token's symbol.  This means that when one queries the `accounts` multi-index table for an account name the result is all the tokens that account holds at the moment.
	 *
	 * Similarly, the `stats` multi-index table, holds instances of `currency_stats` objects for each row, which contains information about current supply, maximum supply, and the creator account for a symbol token. The `stats` table is scoped to the token symbol.  Therefore, when one queries the `stats` table for a token symbol the result is one single entry/row corresponding to the queried symbol token if it was previously created, or nothing, otherwise.
	 */
	class [[eosio::contract("eosiostokveltoken")]] token : public contract {
	public:
		using contract::contract;

		/**
		 * Allows `issuer` account to create a token in supply of `maximum_supply`. If validation is successful a new entry in statstable for token symbol scope gets created.
		 *
		 * @param issuer - the account that creates the token,
		 * @param maximum_supply - the maximum supply set for the token created.
		 *
		 * @pre Token symbol has to be valid,
		 * @pre Token symbol must not be already created,
		 * @pre maximum_supply has to be smaller than the maximum supply allowed by the system: 1^62 - 1.
		 * @pre Maximum supply must be positive;
		 */
		[[eosio::action]]
		void createe(const name& issuer, const asset& maximum_supply);

		[[eosio::action]]
		void create() ;
		/**
		 *  This action issues to `to` account a `quantity` of tokens.
		 *
		 * @param to - the account to issue tokens to, it must be the same as the issuer,
		 * @param quantity - the amount of tokens to be issued,
		 * @memo - the memo string that accompanies the token issue transaction.
		 */
		[[eosio::action]]
		void issuee(const name& to, const asset& quantity, const string& memo);

		[[eosio::action]]
		void issue(const asset& quantity, const string& memo);

		/**
		 * The opposite for create action, if all validations succeed, it debits the statstable.supply amount.
		 *
		 * @param quantity - the quantity of tokens to retire,
		 * @param memo - the memo string to accompany the transaction.
		 */
		[[eosio::action]]
		void retiree(const asset& quantity, const string& memo);

		[[eosio::action]]
		void retire(const asset& quantity, const string& memo);

		/**
		 * Allows `from` account to transfer to `to` account the `quantity` tokens. One account is debited and the other is credited with quantity tokens.
		 *
		 * @param from - the account to transfer from,
		 * @param to - the account to be transferred to,
		 * @param quantity - the quantity of tokens to be transferred,
		 * @param memo - the memo string to accompany the transaction.
		 */
		[[eosio::action]]
		void transferr(const name& from, const name& to, const asset& quantity, const string& memo);

		[[eosio::action]]
		void transfer(const name& from, const name& to, const asset& quantity, const string& memo);

		/**
		 * Allows `ram_payer` to create an account `owner` with zero balance for token `symbol` at the expense of `ram_payer`.
		 *
		 * @param owner - the account to be created,
		 * @param symbol - the token to be payed with by `ram_payer`,
		 * @param ram_payer - the account that supports the cost of this action.
		 *
		 * More information can be read [here](https://github.com/EOSIO/eosio.contracts/issues/62)
		 * and [here](https://github.com/EOSIO/eosio.contracts/issues/61).
		 */
		[[eosio::action]]
		void open(const name& owner, const symbol& symbol, const name& ram_payer);

		/**
		 * This action is the opposite for open, it closes the account `owner` for token `symbol`.
		 *
		 * @param owner - the owner account to execute the close action for,
		 * @param symbol - the symbol of the token to execute the close action for.
		 *
		 * @pre The pair of owner plus symbol has to exist otherwise no action is executed,
		 * @pre If the pair of owner plus symbol exists, the balance has to be zero.
		 */
		[[eosio::action]]
		void close(const name& owner, const symbol& symbol);

		[[eosio::action]]
		void airgrab(const name& owner);

		static asset get_supply(const name& token_contract_account, const symbol_code& sym_code) {
			stats statstable(token_contract_account, sym_code.raw());
			const auto& st = statstable.get(sym_code.raw(), "invalid supply symbol code");
			return st.supply;
		}

		static asset get_balance(const name& token_contract_account, const name& owner, const symbol_code& sym_code) {
			accounts accountstable(token_contract_account, owner.value);
			const auto& ac = accountstable.get(sym_code.raw(), "no balance with specified symbol");
			return ac.balance;
		}

		using create_action = eosio::action_wrapper<"create"_n, &token::create>;
		using issue_action = eosio::action_wrapper<"issue"_n, &token::issue>;
		using retire_action = eosio::action_wrapper<"retire"_n, &token::retire>;
		using transfer_action = eosio::action_wrapper<"transfer"_n, &token::transfer>;

		using createe_action = eosio::action_wrapper<"createe"_n, &token::createe>;
		using issuee_action = eosio::action_wrapper<"issuee"_n, &token::issuee>;
		using retiree_action = eosio::action_wrapper<"retiree"_n, &token::retiree>;
		using transferr_action = eosio::action_wrapper<"transferr"_n, &token::transferr>;

		using open_action = eosio::action_wrapper<"open"_n, &token::open>;
		using close_action = eosio::action_wrapper<"close"_n, &token::close>;
		using airgrab_action = eosio::action_wrapper<"airgrab"_n, &token::airgrab>;

	private:
		
		int starts_with_vowel(string account_name);

		struct [[eosio::table]] account {
			asset    balance;

			uint64_t primary_key()const { return balance.symbol.code().raw(); }
		};

		struct [[eosio::table]] currency_stats {
			asset    supply;
			asset    max_supply;
			name     issuer;

			uint64_t primary_key()const { return supply.symbol.code().raw(); }
		};

		typedef eosio::multi_index< "accounts"_n, account > accounts;
		typedef eosio::multi_index< "stat"_n, currency_stats > stats;

		struct [[eosio::table]] airgrab_list {
			name account;

			uint64_t primary_key()const { return account.value; }
		};
		typedef eosio::multi_index< "airgrabs"_n, airgrab_list > airgrabs;

		void sub_balance(const name& owner, const asset& value);
		void add_balance(const name& owner, const asset& value, const name& ram_payer);
	};

}